Mathlib/Data/Set/Subset.lean

```diff
@@ -0,0 +1,346 @@
+/-
+Copyright (c) 2024 Miguel Marco. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Miguel Marco
+-/
+import Mathlib.Data.Set.Function
+import Mathlib.Data.Set.Functor
+import Mathlib.Lean.Expr.ExtraRecognizers
+
+/-!
+# Subsets as subtypes
+
+This file defines the function `set_restrict` that converts sets in a type, to sets in a subtype
+(as the elements in the subtype whose lift lives in the set), and some notation and simplification
+lemmas for convenience.
+
+More precisely, given `α : Type` and `A : Set α`, `set_restrict A : Set α → Set A` takes
+some `B : Set α` and produces some `Set ↑A` whose elements are `{ x : ↑A | ↑x ∈ B}`.
+
+It also defines the notation `A ↓∩ B` for `set_restrict A B`.
+
+## Implementation
+
+The simplification lemmas are designed to simplify the expression appearing inside the `↓∩`
+operation, if possible.
+
+## Tags
+
+subsets
+-/
+
+open Lean PrettyPrinter Delaborator SubExpr in
+/--
+Sets of a subtype coerced to the ambient type are denoted with `↑`.
+-/
+@[delab app.Set.image]
+def delab_set_image_subtype : Delab := do
+  let #[α, _, f, _] := (← getExpr).getAppArgs | failure
+  guard <| f.isAppOfArity ``Subtype.val 2
+  let some _ := α.coeTypeSet? | failure
+  let e ← withAppArg delab
+  `(↑$e)
+
+open Set
+
+universe u
+
+variable {α β : Type u} {A B C: Set α} {D E : Set ↑A}
+variable {S : Set (Set α)} {T : Set (Set ↑A)} {i : β → Set α} {j : β → Set ↑A}
+
+namespace Subset
+
+/--
+Given two sets `A` and `B`, `set_restrict A B` is the set of `↑A` formed by the elements
+whose value is in `B`.
+-/
+def set_restrict (A B : Set α) : Set ↑A := restrict A B
+
+/--
+`A ↓∩ B` denotes `restrict A B`.
+-/
+infixl:67 " ↓∩ "  => set_restrict
+
+@[simp]
+lemma mem_set_restrict_iff (x : A): x ∈  (A ↓∩ B) ↔ ↑x ∈ B := by rfl
+
+@[simp]
+lemma set_restrict_empty : A ↓∩ (∅ : Set α) = (∅ : Set A) := rfl
+
+@[simp]
+lemma set_restrict_self : A ↓∩ A = univ := by
+  ext x
+  simp only [mem_set_restrict_iff, Subtype.coe_prop, mem_univ]
+
+@[simp]
+lemma set_restrict_univ : A ↓∩ (univ : Set α) = (univ : Set A) := rfl
+
+@[simp]
+lemma set_restrict_union :  A ↓∩  (B ∪ C) = (A ↓∩ B) ∪ (A ↓∩ C) :=  rfl
+
+@[simp]
+lemma set_restrict_inter : A ↓∩ (B ∩ C) = (A ↓∩ B) ∩ (A ↓∩ C) :=  rfl
+
+@[simp]
+lemma set_restrit_compl : A ↓∩ Bᶜ = (A ↓∩ B)ᶜ  := by
+  apply Eq.refl
+
+lemma set_restrict_eq_univ_of_subset (h : A ⊆ B) : A ↓∩ B = univ := by
+  ext x
+  simp only [mem_set_restrict_iff, mem_univ, iff_true]
+  exact h x.2
+
+@[simp]
+lemma restrict_subset_restrict_iff: A ↓∩ B ⊆ A ↓∩ C ↔ A ∩ B ⊆ A ∩ C := by
```

Comment:
```suggestion
lemma restrict_subset_restrict_iff: A ∩ B ⊆ A ∩ C ↔ A ↓∩ B ⊆ A ↓∩ C := by
```
My hunch is that this is the direction I want to go when I use this API, i.e. when I see that everything is intersected by the same set, I want to pass the problem down to the subtype topology. I also don't think this should be a `simp` lemma (in either direction), so that `simp` can be applied while we manually choose if we wish to "stay within the subtype".

Reply:
I see what you mean. I implemented it with a particular use case in mind (which is basically educational: I want this to be used by the students of my general topology course). In that case, I just want `simp` to "simplify the problem as much as possible", and checking if a set is contained in other sounds less painful in the ambient space.

But as you say, it could also make sense to leave it as a non-simp lemma, and maybe add an equivalent simp lemma at the beginning of my exercises for my student's convenience.

Reply:
My own hunch is that this lemma is fine in the current direction, but it's not clear whether it should be a simp lemma.

My thinking is that it's *usually* better to avoid subtypes as much as possible, and so this lemma lets you leave the world of subtypes and go back to the "universe" type.

The reason it might not a good simp lemma is that it changes the types of the objects under consideration, and so `simp` might suddenly make a hypothesis no longer applicable. This isn't a hard "no" to it being a simp lemma, but it's worth thinking about.

Reply:
The direction of the lemma remains to be seen in actual applications, so I don’t have a strong opinion on that.

I am slightly more convinced this shouldn’t be `simp`, because I’d like `simp` to apply all relevant lemmas on the restricted subsets first (e.g. de Morgan), without taking me out of subtype world. I want to be the one to decide, “I’m done with subtype topology now; beam me back up”.

Reply:
Would it make sense to be `aesop` instead of simp?

Reply:
> The direction of the lemma remains to be seen in actual applications, so I don’t have a strong opinion on that.
> 
> I am slightly more convinced this shouldn’t be `simp`, because I’d like `simp` to apply all relevant lemmas on the restricted subsets first (e.g. de Morgan), without taking me out of subtype world. I want to be the one to decide, “I’m done with subtype topology now; beam me back up”.

From that point of view, it fits better in the direction of transforming expressions in the subtype to expressions in the ambient space.